src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -3187,18 +3211,22 @@ void InterpreterSelectQuery::executeDistinct(QueryPlan & query_plan, bool before
 
 
 /// Preliminary LIMIT - is used in every source, if there are several sources, before they are combined.
-void InterpreterSelectQuery::executePreLimit(QueryPlan & query_plan, bool do_not_skip_offset)
+bool InterpreterSelectQuery::executePreLimit(QueryPlan & query_plan, bool do_not_skip_offset)
 {
     auto & query = getSelectQuery();
     /// If there is LIMIT
     if (query.limitLength())
     {
         LimitInfo lim_info = getLimitLengthAndOffset(query, context);
 
+        if (lim_info.fractional_offset > 0 || lim_info.fractional_limit > 0)
+            // Preliminary Limits mustn't be added if there is a fractional limit/offset
```

Comment:
Could you add bit more reasoning in the comment on why prelim should not be applied? It will be useful for future developers.